### PR TITLE
feat(dop,gittar): update preReceiveHook and add create pipeline link to remote

### PR DESCRIPTION
**Pull request by littlejiancc**
#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=287033&issueFilter__urlQuery=eyJzdGF0ZXMiOls0NDA2LDQ0MDddfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&iterationID=1115&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      update push event and add create pipeline link to remote        |
| 🇨🇳 中文    |      修复preReceiveHook问题，remote增加link信息       |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/gittar/helper/git_cmd.go [hunk 1]
```diff
@@ -183,10 +183,15 @@ func RunProcess(service string, c *webcontext.Context) {
 		}
 		logrus.Infof("push header:" + string(header))
 		re := regexp.MustCompile(
-			`(?mi)(?P<before>[0-9a-fA-F]{40}) (?P<after>[0-9a-fA-F]{40}) (?P<ref>refs\/(heads|tags)\/.*?)\0`,
+			`(?mi)(?P<before>[0-9a-fA-F]{40}) (?P<after>[0-9a-fA-F]{40}) (?P<ref>refs\/(heads|tags)\/\S*)`,
 		)
+		// remove 0000
```
Comment (sfwn) [inline modules/gittar/helper/git_cmd.go:188]: you should add some comment or link to explain why remove 0000

and add unit test if could.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/gittar/helper/git_cmd.go (hunk 1 above)
Comment (littlejiancc) [reply modules/gittar/helper/git_cmd.go]: done

File: modules/gittar/helper/git_cmd.go [hunk 2]
```diff
@@ -201,7 +202,16 @@ func RunProcess(service string, c *webcontext.Context) {
...
 		if preReceiveHook(pushEvents, c) {
-
+			// Only when one branch is created will it be written to the writer
+			// Refer to github
+			if len(pushEvents) == 1 && pushEvents[0].IsCreateNewBranch() {
+				// TODO write create pipeline link to writer
+				//c.GetWriter().Write(NewReportStatus(
+				//	"unpack ok",
+				//	"ok "+pushEvents[0].Ref,
+				//	fmt.Sprintf("\nCreate a pipeline request for '%s' on Erda by visiting:\n"+
```
Comment (sfwn) [inline modules/gittar/helper/git_cmd.go:212]: gittar shouldn't create pipeline related link directly.
you should provide a method for invoker to set extra infos, include links.

File: modules/gittar/helper/git_cmd.go [hunk 3]
```diff
@@ -220,6 +230,16 @@ func RunProcess(service string, c *webcontext.Context) {
 	}
 }
 
+// removeEndMarkerFromHeader remove end marker '0000' from header
+// see https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt
```
Comment (sfwn) [inline modules/gittar/helper/git_cmd.go:234]: Can you point out which line it is?

File: modules/gittar/helper/git_cmd_test.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestRemoveEndMarkerFromHeader(t *testing.T) {
+	tt := []struct {
+		header string
+		want   string
+	}{
+		{"00aa0000000000000000000000000000000000000000 5cf4abb264b4a3a0d99e2b967ef8980a1cc41e77 refs/heads/test1 report-status-v2 side-band-64k object-format=sha1 agent=git/2.33.10000",
```
Comment (sfwn) [inline modules/gittar/helper/git_cmd_test.go:28]: it looks ugly.
add new line like below.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/gittar/helper/git_cmd_test.go (hunk 4 above)
Comment (littlejiancc) [reply modules/gittar/helper/git_cmd_test.go:28]: done

File: modules/gittar/helper/git_cmd.go (hunk 2 above)
Comment (littlejiancc) [reply modules/gittar/helper/git_cmd.go:212]: Using the git push command case

File: modules/gittar/helper/git_cmd.go (hunk 3 above)
Comment (littlejiancc) [reply modules/gittar/helper/git_cmd.go]: done

**Review by sfwn — APPROVED** [review]

**Pull request merged**